styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java

```diff
@@ -71,18 +70,29 @@ public void transitionInto(RunState state) {
         // Perform rate limited submission on a separate thread pool to avoid blocking the caller.
         executor.submit(() -> {
           rateLimiter.acquire();
+
+          final RunSpec runSpec;
           try {
-            final String executionId = dockerRunnerStart(state);
-            // this is racy
-            final Event submitted = Event.submitted(state.workflowInstance(), executionId);
-            try {
-              stateManager.receive(submitted);
-            } catch (StateManager.IsClosed isClosed) {
-              LOG.warn("Could not send 'created' event", isClosed);
-            }
+            runSpec = createRunSpec(state);
           } catch (ResourceNotFoundException e) {
             LOG.error("Unable to start docker procedure.", e);
             stateManager.receiveIgnoreClosed(Event.halt(state.workflowInstance()));
+            return;
+          }
+
+          // Emit submitted event first to guarantee it is observed before events from the pod
+          final Event submitted = Event.submitted(state.workflowInstance(), runSpec.executionId());
+          try {
+            stateManager.receive(submitted);
+          } catch (StateManager.IsClosed isClosed) {
```

Comment:
Kind of unsure why we need to check these. `StateManager` closing means we're exiting, right? Not sending events anymore then is normal operation. It would be fine to exit the event handler thread right there with an unchecked exception.